### PR TITLE
Use C#8 and reference type nullability

### PR DESCRIPTION
**Pull request by yarhamjohn**
Team Orca are using C#8 and have enabled C#8 nullability checks so when we build our projects we see lots of nullability warnings from these libraries. Hopefully enabling this feature here will resolve these warnings.

### CODE REVIEW

File: Source/ULibs.TinyJsonDeser/JsonDeserializer.cs [hunk 1]
```diff
@@ -60,7 +60,7 @@ internal sealed class JsonDeserializer
         /// <c>bool</c>, a <c>string</c>, a <c>double</c>, an <c>object[]</c>, an <c>IDictionary&lt;string, object&gt;</c>
         /// or <c>null</c>.</param>
         /// <returns>Whether or not the parse attempt succeeded.</returns>
-        public bool TryParseValue(string json, out object output)
+        public bool TryParseValue(string json, out object? output)
```
Comment (nyctef) [inline Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: Apparently we can make these `Try*` methods slightly better by adding a `[NotNullWhen(returnValue: true)]` attribute to the out parameter? https://www.meziantou.net/csharp-8-nullable-reference-types.htm#conditional-post-con

File: Source/ULibs.TinyJsonDeser/JsonDeserializer.cs [hunk 2]
```diff
@@ -503,7 +503,10 @@ private static bool TryParseArray(string json, ref int offset, out object[] outp
                     SkipWhiteSpace(json, ref index);
                     if (TryParseValue(json, ref index, out var firstElement))
                     {
-                        list.Add(firstElement);
+                        if (firstElement != null)
+                        {
```
Comment (nyctef) [inline Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: ...which hopefully means we won't need these extra `if` checks

File: Source/ULibs.TinyJsonDeser/JsonDeserializer.cs (hunk 1 above)
Comment (yarhamjohn) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: Hmm these attributes don't seem to be recognised

Comment (nyctef) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: Ah yeah, we might not have them here - sounds like we need to either update this project to `netstandard2.1` or write out the attributes manually? https://www.infoq.com/articles/CSharp-8-Framework/ . So long as we reference an attribute with the right name `System.Diagnostics.CodeAnalysis.NotNullWhenAttribute` I think it should work ...

(The latter option might be better since I think that means it'll work for `net472` as well?)

Comment (yarhamjohn) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: So upgrading the projects to netstandard2.1 raises issues with the tests project (which is net4.6.1) and maybe other issues in dependencies?

In terms of referencing the attribute with the right name, this doesn't appear to work - those attributes don't exist on `System.Diagnostics.CodeAnalysis` in netstandard2.0 it appears. The work around in the article is to actually define the attributes manually which feels a bit overkill for this PR?

Comment (nyctef) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: > The work around in the article is to actually define the attributes manually which feels a bit overkill for this PR?

Sorry, yes, that's what I meant. I don't think it's too bad since we'd only actually have to define `NotNullWhenAttribute`, and it would make the `Try*` methods here more usable - it's not a huge deal though so I could go either way.

Comment (yarhamjohn) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: Looks like that would work. However, it turns out that there are some occasions where output can be null and the return value can be true so I don't think this is a solution here. See line 197/198

Comment (nyctef) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: Ah, good spot. I think we need to look into this deeper - I'm guessing the new `if` statement around line 506 is probably a bug (since it'll filter out `null` values from the returned array?) but some of the others might still be necessary?

In `TryParseObjectKeyValuePair` I think we can add the attribute to `TryParseString`, which means that `key` should never be null, but then the value can totally be null so we'd need to change the type of the `KeyValuePair` appropriately.

It looks like there's a similar change in `SerializeGenericDictionary` which I'm now a bit worried about - I think we should be able to serialize a null value there too?

I'm a bit worried the build is still green at this point - we probably need to add some more unit tests.

Comment (nyctef) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: I don't think we should be changing any actual code here :/

Comment (yarhamjohn) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: I agree - I am mostly trying to work out how it fits together...

**Comment by yarhamjohn** [comment]
I _think_ this is better now - it doesn't change any code but deals with the nullability warnings. If you are happy I will squash the commits!

Annoyingly the `[NotNullWhen]` attribute doesn't help the parser understand in which situations it can be null or not so we still need the bang operator to tell it we know it can't be null here.

**Comment by nyctef** [comment]
> Annoyingly the `[NotNullWhen]` attribute doesn't help the parser understand in which situations it can be null or not so we still need the bang operator to tell it we know it can't be null here.

Maybe the attribute actually needs to be declared in the `System.Diagnostics.CodeAnalysis` namespace? Bit of a guess but I think that's how it's worked for similar backports in the past.

File: Source/ULibs.TinyJsonDeser/JsonDeserializer.cs [hunk 3]
```diff
@@ -646,15 +646,34 @@ private static bool TryParseObject(string json, ref int offset, out IDictionary<
...
             return false;
         }
+
+        /// <summary>Specifies that when a method returns <see cref="ReturnValue"/>, the parameter will not be null even if the corresponding type allows it.</summary>
+        [AttributeUsage(AttributeTargets.Parameter, Inherited = false)]
+#if INTERNAL_NULLABLE_ATTRIBUTES
+    internal
+#else
+        public
+#endif
```
Comment (nyctef) [inline Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: Should this just always be `internal`?

**Comment by yarhamjohn** [comment]
> > Annoyingly the `[NotNullWhen]` attribute doesn't help the parser understand in which situations it can be null or not so we still need the bang operator to tell it we know it can't be null here.
> 
> Maybe the attribute actually needs to be declared in the `System.Diagnostics.CodeAnalysis` namespace? Bit of a guess but I think that's how it's worked for similar backports in the past.

Yes that works!

File: Source/ULibs.TinyJsonDeser/JsonDeserializer.cs (hunk 3 above)
Comment (yarhamjohn) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: yup now you mention it I remember reading that in one of the articles you linked

**Comment by nyctef** [comment]
Looking good now - I think there's a few remaining things we need to do:
- get a branch package of this tested in RCC to make sure it does what we want
- probably add a few more unit tests for the things that could have been broken (I think it was stuff like `null` values in arrays being deserialized vs filtered out, stuff like that?)
- get a review from @ChrisLambrou since I think he's done most of the previous work here

**Comment by yarhamjohn** [comment]
Is there a reason (I assume there is) we cannot upgrade these libraries to `netcoreapp3.1`? Will it break other people who depend on it?

**Comment by ChrisLambrou** [comment]
> Is there a reason (I assume there is) we cannot upgrade these libraries to netcoreapp3.1? Will it break other people who depend on it?

They're used by existing .NET Framework code. For now, target the lowest common denominator, which is netstandard2.0. It would be good to run the tests of a variety of different targets, though.

**Comment by yarhamjohn** [comment]
> Looking good now - I think there's a few remaining things we need to do:
> 
> * get a branch package of this tested in RCC to make sure it does what we want
> * probably add a few more unit tests for the things that could have been broken (I think it was stuff like `null` values in arrays being deserialized vs filtered out, stuff like that?)
> * get a review from @ChrisLambrou since I think he's done most of the previous work here

I have tested this branch in RCC here: https://github.com/red-gate/RedgateChangeControl/pull/1763 and it seems to be an improvement, though the issues around netcoreapp3.1 having different nullability rules to netstandard2.0 cannot currently be avoided.

I have added a few more test cases.

File: Source/ULibs.TinyJsonDeser/JsonDeserializer.cs (hunk 1 above)
Comment (ChrisLambrou) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: There's a fundamental problem here. You're trying to get this code to say it'll never return `null`, when that's a very explicitly valid case. Look at the XML doc comments. Line 61.

File: Source/ULibs.TinyJsonDeser/JsonDeserializer.cs [hunk 4]
```diff
@@ -144,7 +144,7 @@ public bool TryParseString(string json, out string output)
         /// <param name="json">The json string to parse.</param>
         /// <param name="output">If the parse succeeded, this will hold the parsed result.</param>
         /// <returns>Whether or not the parse attempt succeeded.</returns>
-        public bool TryParseArray(string json, out object[] output)
+        public bool TryParseArray(string json, out object?[]? output)
```
Comment (ChrisLambrou) [inline Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: Again, you're trying to state that the objects in the array will never be `null`, but that's simply not true. This library parses json, and `null` is a valid json type that can be returned. We could modify the library to return something like `JTokens` rather than built-in types, but that's a significant breaking change, and I'm not sure there's any call for it.

File: Source/ULibs.TinyJsonDeser/JsonDeserializer.cs [hunk 5]
```diff
@@ -161,7 +161,7 @@ public bool TryParseArray(string json, out object[] output)
         /// <param name="json">The json string to parse.</param>
         /// <param name="output">If the parse succeeded, this will hold the parsed result.</param>
         /// <returns>Whether or not the parse attempt succeeded.</returns>
-        public bool TryParseObject(string json, out IDictionary<string, object> output)
+        public bool TryParseObject(string json, out IDictionary<string, object?>? output)
```
Comment (ChrisLambrou) [inline Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: Same. You're saying the object in the dictionary can't be `null`, but it can, and deliberately so. No wonder you're struggling with this PR!

**Comment by ChrisLambrou** [comment]
I added some comments, and got confused. I've deleted them. Ignore my stupidity.

**Comment by nyctef** [comment]
> though the issues around netcoreapp3.1 having different nullability rules to netstandard2.0 cannot currently be avoided.

I guess we could potentially spam a lot of `#if NETCOREAPP3_1` everywhere? Probably not worth the additional complexity, though, unless we actually decide we want to get to zero warnings

Looks good to me otherwise 👍

**Comment by ChrisLambrou** [comment]
You can in theory have multiple versions of the source file, for different target runtimes. That's complicate the NuGet packaging a bit, but I think it's doable. It depends on how much effort you want to put in, and whether or not you want to put it in right now.

**Comment by ChrisLambrou** [comment]
Do we need to consider putting `#nullable enable` at the top of these files? Whilst we can require C# 8.0 language support, there's no guarantee that every consuming library will have nullability support enabled. Or would that actually make things worse, and leaving it off is safer? I can't remember!

**Comment by yarhamjohn** [comment]
I reckon this is a decent starting point - it reduces the warnings considerably but would be worth investigating the different runtimes issue later

**Comment by yarhamjohn** [comment]
> #nullable enable

I am not sure tbh - nullability here is enabled at the project level so would adding this at the file level actually change any behaviour? If consumers not using nullability can't see the project-level enablement, then it seems like we would need it at the file level otherwise the nullable types will not be understood by the compiler?

**Comment by ChrisLambrou** [comment]
We don’t have nullability enabled in SQL Monitor yet, and some files have started to selectively enable it using that directive. It definitely changes the behaviour. I don’t know if it’s in the right direction, though. Let me do some testing.

**Comment by yarhamjohn** [comment]
Thanks for the update Chris. That's good to know - I will add this to the PR!

Thanks for your help with this

**Comment by yarhamjohn** [comment]
@ChrisLambrou That is hopefully fixed now - are you happy for me to merge?

**Comment by ChrisLambrou** [comment]
The build is failing for reasons relating to the language version.

```
error CS1617: Invalid option '8' for /langversion. Use '/langversion:?' to list supported values.
```

It's using `C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\Roslyn\csc.exe`. Is that the right version? Maybe we need to change the agent requirements for the build configuration.

File: Source/ULibs.TinyJsonDeser/JsonDeserializer.cs [hunk 6]
```diff
@@ -144,7 +145,7 @@ public bool TryParseString(string json, out string output)
         /// <param name="json">The json string to parse.</param>
         /// <param name="output">If the parse succeeded, this will hold the parsed result.</param>
         /// <returns>Whether or not the parse attempt succeeded.</returns>
-        public bool TryParseArray(string json, out object[] output)
+        public bool TryParseArray(string json, out object?[]? output)
```
Comment (ChrisLambrou) [inline Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: Shouldn't this method, and all of the other TryParse methods, also have the `[NotNullWhen(true)]` attribute?

Comment (yarhamjohn) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: So I initially tried that but it turns out it is possible for `TryParseValue` to set `output = null` when returning `true` (see line 199). This means its callers also can't be sure, though really this only matters to its overload.That said, I have found a couple of places where that wasn't an issue and have added them

Comment (ChrisLambrou) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: Oh, that's worrying. As far as I can remember, all of the TryParse methods should produce a non-null result if the parse succeeds, with a mere handful of notable exceptions (e.g. TryParseValue, where `null` is a legitimate JSON value). If that's not true, there might be bugs.

Comment (ChrisLambrou) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: I'm struggling to remember, though. It's a long time since I wrote this.

Comment (yarhamjohn) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: I will take a closer look at the TryParseNull method

**Comment by yarhamjohn** [comment]
> The build is failing for reasons relating to the language version.
> 
> ```
> error CS1617: Invalid option '8' for /langversion. Use '/langversion:?' to list supported values.
> ```
> 
> It's using `C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\Roslyn\csc.exe`. Is that the right version? Maybe we need to change the agent requirements for the build configuration.

Ah yes looks like its a flaky problem so the agents probably need revision

File: Source/ULibs.TinyJsonDeser/JsonDeserializer.cs (hunk 6 above)
Comment (ChrisLambrou) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: I should probably checkout the code and have a proper look, rather than just commenting on the changes.

Comment (ChrisLambrou) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: It occurs to me that we could avoid the need for the NullabilityAttribute hack altogether by returning non-null sentinel values for all of the parse methods for reference types. Something like this:

```c#
public static bool TryParseString(string json, ref int offset, out string value)
{
    if (<can-parse>)
    {
        value = <parsed-value>;
        return true;
    }
    value = "";  / Sentinel value to ensure we always output a non-null value.
    return false;
}

public static bool TryParseArray(string json, ref int offset, out object?[] value)
{
    if (<can-parse>)
    {
        value = <parsed-value>;
        return true;
    }
    value = new object[0];  // Sentinel value to ensure we always output a non-null value.
    return false;
}
```

What do you think?

Comment (yarhamjohn) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: That feels like it would be nicer - its already done in a few of the parsing methods

Comment (ChrisLambrou) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: I've tried the above on a separate branch. Here's the PR, so you can see the changes vs the original code on master. I'm not poposing it be merged. It's just an illustration of an alternative approach. If we're happy with it, we can merge that into this branch. If not, delete it.

Comment (nyctef) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: To me it feels more correct/ergonomic to return null when a `TryParse` function fails, since it's a genuine lack of a value and users might accidentally use an incorrect value otherwise?

Comment (ChrisLambrou) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: Shall I bring that change into this branch, then? I might have to have a look at any changes neccessary in the tests, first.

Comment (ChrisLambrou) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: Actually, that's nonsense. The tests don't care about nullability. Maybe they should?

Comment (nyctef) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: Yeah, more tests around null values would probably be good - I think we hit a few cases earlier where we changed null behavior but the tests didn't break :) John added a few extra test cases, but more is always good...

**Comment by yarhamjohn** [comment]
> > The build is failing for reasons relating to the language version.
> > ```
> > error CS1617: Invalid option '8' for /langversion. Use '/langversion:?' to list supported values.
> > ```
> > 
> > 
> > It's using `C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\Roslyn\csc.exe`. Is that the right version? Maybe we need to change the agent requirements for the build configuration.
> 
> Ah yes looks like its a flaky problem so the agents probably need revision

For some reason neither the master branch nor any of the branches run successfully on Team City now - presumably there has been a configuration change on the agents in the last week or so as this branch was working previously, as was master when run on agents with `VISUALSTUDIO_VERSION = 2019`

image

File: Source/ULibs.TinyJsonDeser/JsonDeserializer.cs (hunk 6 above)
Comment (ChrisLambrou) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: I meant enabling null reference types in the test project.It's currently at C# 7.3 with null reference types disabled.

Comment (yarhamjohn) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: I think I had ignored it because changing the language meant changing the target framework and the test projects have old-style csproj and it all felt like too much work for debatable benefit - however I may have misunderstood what was needed

Comment (ChrisLambrou) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: Okay, I've merged the change that uses non-null default values for when the TryParse methods fail.

Comment (ChrisLambrou) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: > To me it feels more correct/ergonomic to return null when a TryParse function fails, since it's a genuine lack of a value and users might accidentally use an incorrect value otherwise?

Sorry. I missed this, I think in a mid-air collision. I only just saw it when I hit refresh.

I'm not sure I agree that it's more correct or ergonomic to return null. What the contract really wants to express is that the output value is undefined when the parse fails, but the language and type system isn't rich enough to really support that yet and we have to set the output parameter to something. But there's nothing special about null here. Returning null would still be defined behaviour (we'd be defining it), and users might still accidentally try to use it without checking the return value first.

The emerging fix to this scenario is the `NotNullWhenAttribute` (the real one, not our fake), but it's still a kludge. It doesn't truly give us type safety in the language. It doesn't even give us a compilation error by default, either. Just a warning.

It'd be nice if there was a ubiquitous `Option<T>` type we could rely on, so callers genuinely only get access to the value if the parse succeeds, but sadly there isn't yet. And C# devs tend to find that sort of thing fugly, still, though it's been getting progressively better in recent years.

One final potential problem with taking the fake attribute approach. We'd be defining a `System.Diagnostics.CodeAnalysis.NotNullWhenAttribute` type in the consuming project. If the consumer tries to make use of two or more libraries that do this, the names would conflict, and boom! It's not a sustainable pattern for us to follow.

As I've typed this comment, I've been somewhat on the fence, but that last issue is a clincher for me. I think it's asking for trouble to declare a polyfill for a system type and expose it to users like that. I think it's better to accept the risk that users might mess up when trying to use a library that follows a fairly idiomatic pattern in .NET. Too bad for them if they get it wrong, but I think the risk is low _(and egg on their faces, not ours, if they do go wrong!)_.

Comment (nyctef) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: > _(and egg on their faces, not ours, if they do go wrong!)_.

I don't think that's a particularly good attitude for library development - we should all be about the "pit of success" here 😛 

I totally agree with the rest of your arguments, though - I guess if we wanted to do the attribute polyfill properly we'd need a whole other microlibrary which consumers could independently import?

**Comment by yarhamjohn** [comment]
Think we are good to go with this now?

File: Source/ULibs.TinyJsonDeser/JsonDeserializer.cs (hunk 6 above)
Comment (ChrisLambrou) [reply Source/ULibs.TinyJsonDeser/JsonDeserializer.cs]: > I don't think that's a particularly good attitude for library development - we should all be about the "pit of success" here 😛

Sorry, I was kinda joking. More seriously, I agree, but the only real options for making this any better either involve a significantly breaking API change (e.g return an option type) or the attribute hack that's potentially dangerous. And TryParse is such a common idiom in .NET that the current code feels pretty reasonable and low risk. I've been exploring ways to have multiple versions of this file, so that where the `WhereNotNull` is available, we use it, but when  it's not we fall back to the current implementation. I managed to get it working, but I think it's best to leave it for another PR.

> I totally agree with the rest of your arguments, though - I guess if we wanted to do the attribute polyfill properly we'd need a whole other microlibrary which consumers could independently import?

Yeah. We could possibly do that as a conditional dependency. I'll have a play with that, too.

**Comment by ChrisLambrou** [comment]
Should we squash everything down to a single commit?

**Comment by yarhamjohn** [comment]
Yeah happy to do that - will be a bit cleaner!
